Documentation/botr/readytorun-format.md

```diff
@@ -1,15 +1,77 @@
+
+* In single-input R2R PE files compiled in the large version bubble mode, the metadata manifest contains the set of assemblies containing functions called from the input assembly (possibly with some transitivity due to inlining); in composite R2R PE files the manifest metadata is supposed to contain all assemblies contained within the composite file. The linear indices of the assemblies in the manifest metadata AssemblyRef table correspond to indexing within the `READYTORUN_SECTION_ASSEMBLIES` and `READYTORUN_SECTION_SIMPLE_ASSEMBLY_NAME_LOOKUP` tables.
+
+* In single-input R2R PE files the R2R section `READYTORUN_SECTION_AVAILABLE_TYPES` encodes a native hashtable of types within a given assembly. In composite R2R files, the available types need to be split per input assembly and are referenced from `READYTORUN_SECTION_ASSEMBLIES`.
+
+* In single-input R2R PE files the R2R section `READYTORUN_SECTION_METHOD_ENTRYPOINTS` encodes a native hashtable of methods within a given assembly. In composite R2R files, the method entrypoints need to be split per input assembly and are referenced from `READYTORUN_SECTION_ASSEMBLIES`.
+
+* In single-input R2R PE files the  generic instantiations generated for the input MSIL assembly are placed in `READYTORUN_SECTION_INSTANCE_METHOD_ENTRYPOINTS`. In composite R2R files, this section represents all instance entrypoints emitted within the composite build (i.e. generic instantiations needed by any of the input assemblies). With the proposed change to make the "input MSIL" represent the manifest metadata, we should be able to easily refer to all required reference assemblies in the signature encodings even without the manifest module override (**note:** this is a minute technical detail).
+
+### Required CoreCLR runtime changes
+
+CoreCLR runtime will need to understand the `READYTORUN_FLAG_COMPOSITE` flag and behave accordingly:
+
+* For composite files, we shouldn't be checking MVID's of reference assemblies within the tight version bubble represented by the single file.
+* We need to improve generic instantiation lookup algorithm to be able to locate all instantiations emitted into the composite file.
+* CoreCLR runtime will need to parse the manifest metadata (and possibly the `READYTORUN_SECTION_SIMPLE_ASSEMBLY_NAME_LOOKUP` section) and set up lookup structures to facilitate fast assembly lookup for available types and method entrypoints.
+* The runtime will need to start consulting the `READYTORUN_SECTION_ASSEMBLIES` table to locate MSIL metadata within the composite executable - this logic should definitely take precedence to arbitrary files on the disk to satisfy consistency guarantees.
 
 # PE Headers and CLI Headers
```

Comment:
For the precompiled code, I believe that should get merged automatically by the compiler. For the MSIL it's somewhat more tricky - I think it should be no problem to put all MSIL metadata in the same section - in fact it's probably the logical thing to do - but it's not immediately obvious whether it will automatically fix the entire problem - e.g. if PerfView expects the entire MSIL section to be a single MSIL metadata block, it still wouldn't work.